client/auth/login.jsx

```diff
@@ -103,27 +101,58 @@ module.exports = React.createClass( {
 	},
 
 	canSubmitForm: function() {
+		const auth = this.state.auth;
+
 		// No submission until the ajax has finished
-		if ( this.state.inProgress ) {
+		if ( auth.inProgress ) {
 			return false;
 		}
 
 		// If we have 2fa set then don't allow submission until a code is entered
-		if ( this.state.requires2fa ) {
+		if ( auth.required2faType ) {
 			return parseInt( this.state.auth_code, 10 ) > 0;
 		}
 
 		// Don't allow submission until username+password is entered
 		return this.hasLoginDetails();
 	},
 
-	toggleSelfHostedInstructions: function () {
-	    var isShowing = !this.state.showInstructions;
-	    this.setState( { showInstructions: isShowing } );
+	toggleSelfHostedInstructions: function() {
+		const isShowing = ! this.state.showInstructions;
+		this.setState( { showInstructions: isShowing } );
+	},
+
+	useAuthCode: function() {
+		AuthActions.useAuthCode(
+			this.translate( 'Please enter the verification code generated by your Authenticator mobile application.' )
+		);
+	},
+
+	focusWhenIdle: function( ref ) {
+		if ( this.state.auth.inProgress === false ) {
+			ref.focus();
+		}
+	},
+
+	updateLogin: function( e ) {
+		this.setState( { login: e.target.value } );
+	},
+
+	updatePassword: function( e ) {
+		this.setState( { password: e.target.value } );
+	},
+
+	updateAuthCode: function( e ) {
+		this.setState( { auth_code: e.target.value } );
 	},
```

Comment:
> this will generate the resulting functions in the render method and will always re-render

true, but is it a problem? we've got three closures being generated in a login-type form which is only going to be updating when a user presses a key or an async fetch comes back in.

given the tradeoffs, I'd rather see code that feels better/more suited to the task and _maybe_ slow than code that feels less right and _probably_ faster. that is, if it becomes a problem we could address it but let's not assume it's a problem when it's probably not.

that being said, I welcome discussion on whether or not you find it better/more suited to the problem!